src/main/scala/com/target/data_validator/Emailer.scala

```diff
@@ -108,16 +108,29 @@ object Emailer extends LazyLogging {
 
   def sendMessage(message: Message, body: String, mime: String): Boolean = {
     message.setContent(body, mime)
+    val id = message.hashCode().toHexString
     try {
+      logger.info(s"Sending email #$id [${message.getSubject}] to [${message.getAllRecipients.mkString(", ")}]")
       Transport.send(message)
+      logger.info(s"Email #$id sent totally successfully.")
       false
     }
     catch {
       case sfe: SendFailedException =>
-        logger.error(s"Failed to send message, $sfe")
+        logger.warn(s"Failure to send email #$id: $sfe")
+        if (sfe.getValidSentAddresses.nonEmpty) {
```

Comment:
From the api docs, it appears that all 3 arrays (ValidSent, ValidUnsent and Invalid) will either contain at least one address, or will be null i.e it cannot be an empty array.

Reply:
Yes, this is true.  Although I did test each of the paths and only the first case (`getValidSentAddresses`) threw the NPE.  In my test, the ValidUnsent case did return an empty array. 🤷‍♂️ 

Since this is from a java lib and can return null, it is probably more appropriate to wrap the other calls in an Option. Something like:

```scala
Option(sfe.getValidUnsentAddresses) match {
  case Some(addresses) => logger.error(s"Email #$id was not sent to [${addresses.mkString(", ")}]")
  case None =>
}
Option(sfe.getInvalidAddresses) match {
  case Some(addresses) => logger.error(s"Email #$id has invalid addresses: [${addresses.mkString(", ")}]")
  case None =>
}
```

Reply:
This is the direction I'm going to go when I get back at it this week.